app/screens/channel_files/channel_files.tsx

```diff
@@ -0,0 +1,167 @@
+import Loading from '@components/loading';
+import Search from '@components/search';
+import {General} from '@constants';
+import {useServerUrl} from '@context/server';
+import {useTheme} from '@context/theme';
+import {FileFilter, FileFilters, filterFileExtensions} from '@utils/file';
+import {changeOpacity, getKeyboardAppearanceFromTheme} from '@utils/theme';
+
+import Header from './header';
+
+import type ChannelModel from '@typings/database/models/servers/channel';
+
+const TEST_ID = 'channel_files';
+
+type Props = {
+    channel: ChannelModel;
+    canDownloadFiles: boolean;
+    publicLinkEnabled: boolean;
+}
+
+const edges: Edge[] = ['bottom', 'left', 'right'];
+
+const styles = StyleSheet.create({
+    flex: {
+        flex: 1,
+    },
+    empty: {
+        alignItems: 'center',
+        flex: 1,
+        justifyContent: 'center',
+    },
+    list: {
+        paddingVertical: 8,
+    },
+    loading: {
+        justifyContent: 'center',
+        flex: 1,
+    },
+    searchBar: {
+        marginLeft: 20,
+        marginRight: Platform.select({ios: 12, default: 20}),
+        marginTop: 20,
+    },
+});
+
+const getSearchParams = (channelId: string, searchTerm?: string, filterValue?: FileFilter) => {
+    const term = `channel:${channelId}`;
+    const fileExtensions = filterFileExtensions(filterValue);
+    const extensionTerms = fileExtensions ? ' ' + fileExtensions : '';
+    const searchTerms = searchTerm ? ' ' + searchTerm : '';
+    return {
+        terms: term + searchTerms + extensionTerms,
+        is_or_search: true,
+    };
+};
+
+const emptyFileResults: FileInfo[] = [];
+
+function ChannelFiles({
+    channel,
+    canDownloadFiles,
+    publicLinkEnabled,
+}: Props) {
+    const theme = useTheme();
+    const serverUrl = useServerUrl();
+    const {formatMessage} = useIntl();
+    const searchTimeoutId = useRef<NodeJS.Timeout | null>(null);
+    const [filter, setFilter] = useState<FileFilter>(FileFilters.ALL);
+    const [loading, setLoading] = useState(true);
+    const [term, setTerm] = useState('');
+
+    const [fileInfos, setFileInfos] = useState<FileInfo[]>(emptyFileResults);
+
+    const handleSearch = useCallback(async (searchTerm: string, ftr: FileFilter) => {
+        const searchParams = getSearchParams(channel.id, searchTerm, ftr);
+        const {files} = await searchFiles(serverUrl, channel.teamId, searchParams);
+        setFileInfos(files?.length ? files : emptyFileResults);
+        setLoading(false);
+    }, [serverUrl, channel]);
+
+    useEffect(() => {
+        if (term) {
+            if (searchTimeoutId.current) {
+                clearTimeout(searchTimeoutId.current);
+            }
+
+            searchTimeoutId.current = setTimeout(() => {
+                handleSearch(term, filter);
+            }, General.SEARCH_TIMEOUT_MILLISECONDS);
+        } else {
+            handleSearch(term, filter);
+        }
+    }, [filter, term, handleSearch]);
```

Comment:
This is going to be fun :smile:

Image you write something, and fast you delete it. For example, `a`.
The moment you write `a`, this effect triggers. Since you have  a term (`a`), it will go through the first part of the if, setting a timeout that will eventually execute search (lets call it `search a`).
But then we delete `a`, so this effect will trigger again, and will go through the else branch, calling `handleSearch` directly (lets call it `search all`).

`search all` will execute first, and if everything happens as expected on the network, it will return first, and will set the file infos and loading to the correct values.
But the timeout for `search a` is still running, and will execute, creating a new network call, and eventually, updating again the file infos.

So the user experience will be:
- I write `a`
- I delete the `a`
- I see the loading spinner
- I see the correct results
- Suddenly, the correct results transform into a subset of the results

The simplest solution is to remove the if/else at all. You are doing practically the same on both branches.

That will still give some problems if the network calls return in a different order, but that is way less common in this scenario. Just for fun, let me show you the needed steps for that to happen:
- I write `a`
- I wait for `SEARCH_TIMEOUT_MILLISECONDS` so `handleSearch` is executed (`search a`)
- I write `ab`
- I wait for `SEARCH_TIMEOUT_MILLISECONDS` so `handleSearch` is executed (`search ab`)
- `search a` for whatever reason is taking a lot of time
- `search ab`, on the other hand, returns faster, and sets the file infos
- `search a` finally returns, and set also the file infos.
- The user experience becomes "I had the right results, and now I have more results than I should have"

If we were to address this, the best way would be to store the `searchFiles` request in a reference, and cancel it, but we still don't have support for that. The next best way to address that is to store a timestamp like `lastSearchRequest`. That way, handle search can do something like:
```
const t = Date.now()
lastSearchRequest.current = t;
const {files} = await searchFiles(...);
if (lastSearchRequest.current !== t) {
    return;
}
...
```